mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,964 @@
+    // Number of cache lines accessed by this loop at index i
+    unsigned numberCacheLinesAccessed =
+        cacheLineCount[&loopNest[perm[i]]];
+    if (numberCacheLinesAccessed > maxCacheLinesAccessed)
+      maxCacheLinesAccessed = numberCacheLinesAccessed;
+
+    auxiliaryCost += numberCacheLinesAccessed * iterSubSpaceSize;
+    iterSubSpaceSize *= loopIterCounts[perm[i]];
+  }
+
+  // The optimal permutation is one in which the affine.for ops are arranged in
+  // descending order of their individual cache line access counts from left to
+  // right.
+
+  // However such a permutation will have maximum 'auxiliaryCost' value. But we
+  // want a minimum cost for such a permutation.
+
+  // To reverse the effect, we subtract the auxiliaryCost value from a sentinel
+  // value(maximum possible spatial locality cost). We define sentinel value as
+  // follows: sentinel = Sum_for_all_loops(maxCacheLineAccessed *
+  // iterSubSpaceSize of each loop).
+
+  long double sentinel = 0;
+  iterSubSpaceSize = 1;
+  for (unsigned iterCount : loopIterCounts) {
+    sentinel += maxCacheLinesAccessed * iterSubSpaceSize;
+    iterSubSpaceSize *= iterCount;
+  }
+  return sentinel - auxiliaryCost;
+}
+
+/// Returns the optimal permutation considering both - the locality cost as well
+/// as the parallelism cost. The permutation which has the lowest overall cost is
+/// returned. Returns size-0 permutation in case no valid interchange exists.
+SmallVector<unsigned, 4> LoopInterchange::getBestPermutation(
+    SmallVector<AffineForOp, 4> &loopNest,
+    DenseMap<Value, unsigned> &indVarIdMap) {
+
+  SmallVector<unsigned, 4> bestPerm;
+  SmallVector<Operation *, 8> loadAndStoreOps;
+  long double minCost = LONG_MAX;
+
+  // Get all affine.load and affine.store ops.
+  getAllLoadStores(loopNest[0], loadAndStoreOps);
+
+  // Get size of elements in each memref access. Later to be used
+  // in building reference groups.
+  getElementsSize(loopNest[0]);
+
+  // Get affine access matrices for all load/stores.
+  getAffineAccessMatrices(loopNest[0], loadAndStoreOps, indVarIdMap,
+                          loopNest.size());
+
+  // Get loop carried dependency vector for all affine.for ops in the loop nest.
+  // This is useful in calculating the parallelism cost for each permutation.
+  getAffineLoopCarriedDependencies(loopNest[0], loadAndStoreOps,
+                                   loopNest.size());
+
+  // Get locality information for each affine.forop in the loop nest. This will
+  // be useful in calculating spatial locality cost for each permutation.
+  getCacheLineAccessCounts(loopNest, loadAndStoreOps);
+
+  // Get all permutations which do not violate any dependency constraints.
+  SmallVector<SmallVector<unsigned, 4>, 8> validPerms;
+  getAllValidPermutations(loopNest[0], loopNest.size(),
+                          validPerms);
+
+  // Return size 0 SmallVector to indicate that no interchange exists.
+  if (validPerms.size() < 2)
+    return bestPerm;
```

Comment:
Use an output argument on the function instead of returning SmallVector's which leads to too many copies in general.

Reply:
Done.